src/software/new_geom/convex_polygon.h

```diff
@@ -10,13 +10,30 @@ class ConvexPolygon : public Polygon, public ConvexShape
 {
    public:
     ConvexPolygon() = delete;
+
     /**
      * Returns the area of this convex polygon
      *
      * @return The area of this convex polygon
      */
     double area() const override;
 
+    /**
+     * Returns the centroid of this convex polygon
+     *
+     * @return The centroid of this convex polygon
+     */
+    Point centroid() const;
+
+    /**
+     * Returns the convex polygon expanded towards v
+     * The convex polygon is split in half perpendicular to v and points on the half that
+     * v is pointing are translated by v
+     *
+     * @return a convex polygon expanded by v
+     */
+    ConvexPolygon expand(const Vector& v) const;
```

Comment:
Since this is implemented in `ConvexPolygon`, other "concrete shapes" we have classes for like `Rectangle` also have this function.

Currently (at least IMO) rectangles are not defined with the concept of rotation, ie. their edges always lie along the x or y axis.

The expansion function would allow rectangles to be expanded diagonally to form convex polygons, but what may no longer be considered a rectangle because it ends up rotated / skewed etc.

Edit: realized this is not a virtual function so is `ConvexPolygon` only. This raises a new question: Seems like we may move to replace `Rectangle::expand` with this function because it's more useful. Do we even have a good use-case for a specific `Rectangle` class anymore? We definitely might just pointing this out.

Reply:
Rectangles are useful because they are an abstraction around construction (give 2 points instead of 4 and guaranteed shape) and for accessing the corners (e.g. negxnegy...). It raises a good point: we don't need `Triangle` anymore now that we have `centroid`

Reply:
I don't mind having `Triangle` because it's a very "unconstrained" version of a `ConvexPolygon` (ie. any 3 points make a triangle basically). I think it helps readability at the very least.

On the other hand, `Rectangle` is more constrained in that the instant you skew it, it's technically no longer a rectangle. I agree the interface with the corners for creation and stuff is useful.

Just off the top of my head though, it seems like the lines between `Rectangles` and `ConvexPolygons` get blurred. We represent some parts of the field (the entire boundary, etc.) with traditional `Rectangles`, but now would want to represent other parts like the defense areas with `ConvexPolygons` because they have a more convenient `expand` interface for how they are used (ie. in avoid areas).

just feels weird to have shapes and their usages intermingled like that, but "feels" is not a very strong argument for anything.

Would definitely like opinions @jonathanlew @garethellis0 @akhilveeraghanta @jsn0716 

Possible solutions:
* Remove `Rectangle` and just use `ConvexPolygon`
* Make `expand` virtual and have `Rectangle` override it. Same interface but clamps vectors to one of the 4 cardinal directions
* Do not make `expand` virtual but implement a separate version for rectangles that takes an enum for the direction and the magnitude. eg. `expand(NEGATIVE_X, 1.2)`

Regardless, I do like the changes in this PR.

Reply:
Breaking the invariants for `Rectangle` is a really good point. Here's an idea: for a given vector, break it up into `x` and `y` components and `expand` one after the other. For example, if `v = (-1,2)` then first `expand` with `v = (-1,0)` then `expand` with `v = (0,2)`. That should maintain the invariants of `Rectangle`

Reply:
I don't really see the issue with this function as is? As long as it has no side effects, it's returning a `ConvexPolygon`, so we never create an invalid `Rectangle`?

Reply:
That's true, there's nothing wrong with the function as is, but `expand` is currently used primarily by `Rectangle` so I'll write `Rectangle expand(void)` with the idea above

Reply:
Yea @garethellis0 this was not an issue with the current function as-is, more of a note that we have different use-cases and different `expand` functions for the different geometry types.